### ISSUE TITLE
Convenience method for repeated FilePath.parent()

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3588]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#3588|
|Type|enhancement|
|Created|2008-12-22 03:32:37Z|
|Branch|https://github.com/twisted/twisted/tree/parents-3588|

It would be nice if `FilePath` let you do something like:

  ```python
path = FilePath(__file__).parent(3)
```

where the above was equivalent to:

  ```python
path = FilePath(__file__).parent().parent().parent()
```

Having a differently named function would be OK too.

Attachments:

* [parents.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1a8/1a82834c2bb20a89053a0d5f5c3efe918ef04f7d/881808a09f4e3abe46941a3a0e32c4b3cd1cf486.patch) (2247 bytes) - added by glyph on 2008-12-24 09:06:10Z - A patch which implements this, since it's pretty simple.

<details><summary>Searchable metadata</summary>

```
trac-id__3588 3588
type__enhancement enhancement
reporter__jml jml
priority__normal normal
milestone__ 
branch__branches_parents_3588 branches/parents-3588
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1229916757000000 1229916757000000
changetime__1232267806000000 1232267806000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

How about FilePath(__file__).parents()[3]?

This could also address [#3589](https://github.com/twisted/twisted/issues/3589) with a convenient idiom rather than a specific method: `if possibleParent in possibleChild.parents()`...

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml commented|
|-|-|

I could _definitely_ use that. Your suggestion implies that FilePath(x).parents()[0] === FilePath(x), which is probably the right API but a little confusing given the name.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

Hmm.  I think I'd actually prefer it if `FilePath(x).parents()[0] == FilePath(x).parent()`.  This mirrors the behavior of `segmentsFrom`, too, in your request for `isParent` (you don't want `isParentOrSelf()`, right?)  Why do you say it's the right behavior to include the path itself?

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_4"></a>|@jml commented|
|-|-|

Replying to [glyph](#note_3):
> Hmm.  I think I'd actually prefer it if `FilePath(x).parents()[0] == FilePath(x).parent()`.  This mirrors the behavior of `segmentsFrom`, too, in your request for `isParent` (you don't want `isParentOrSelf()`, right?)  Why do you say it's the right behavior to include the path itself?

Because in the cases where I want to use it, I actually want to iterate over the path and all of its ancestors -- I'm looking for something.

jml

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

Replying to [jml](#note_4):

> Because in the cases where I want to use it, I actually want to iterate over the path and all of its ancestors -- I'm looking for something.

I think in those cases, the extra verbosity of "`chain([it], it.parents())`" is useful for clarity's sake.

It occurs to me that `list(it.parents())[2]` is also pretty easy too; it can easily be a generator.

To strengthen my argument on these points I shall provide a patch.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph removed owner|
|-|-|

Merry Christmas.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_7"></a>|@jml set owner to @glyph|
|-|-|

Hey glyph,

The patch looks good -- thanks for getting to it so quickly. I've got a couple of minor comments:

 * Isn't `assertEqual` preferred over `assertEquals`?
 * I think the test would be clearer if it used more literals and less algorithm.

Thanks,
jml

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

(In [25740]) Branching to 'parents-3588'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

Replying to [jml](#note_7):
> Hey glyph,
>
> The patch looks good -- thanks for getting to it so quickly. I've got a couple of minor comments:

>  * Isn't `assertEqual` preferred over `assertEquals`?

If you say so.  I don't remember anything like that.  Is this written down anywhere?

>  * I think the test would be clearer if it used more literals and less algorithm.

This is pretty vague, but I do think it's pretty crappy that all the variables are one letter.  So let me fix that.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

I hope I got the gist of "more literals"; I couldn't really use "less algorithm", having previously used precisely the correct amount :).

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_12"></a>|@jml set owner to @glyph|
|-|-|

Looks good to me. Please merge.

I realize now that my comment was too cryptic -- sorry. Here's what I meant. Instead of:

```python
def test_parents(self):
    L = []
    pathobj = self.path.child("a").child("b").child("c")
    fullpath = pathobj.path
    lastpath = fullpath
    thispath = os.path.dirname(fullpath)
    while lastpath != self.root.path:
        L.append(thispath)
        lastpath = thispath
        thispath = os.path.dirname(thispath)
    self.assertEqual([x.path for x in pathobj.parents()], L)
```

Try something like:

```python
def test_parents(self):
    pathobj = self.path.child("a").child("b").child("c")
    expectedParents = [
        self.path.child('a'),
        self.path.child('a').child('b'),
        ]
    self.assertEqual(
        [x.path for x in pathobj.parents()],
        [x.path for x in expectedParents])
```

I think the second example makes the expected behaviour of the method clearer. That said, it's your call, and I'm happy for you to merge your patch as is.

Thanks, and sorry about the roundtripping.

jml

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_13"></a>|@jml commented|
|-|-|

Also, regarding `assertEqual` vs `assertEquals`, I don't know whether one is preferred over the other -- that's why I asked the question. I have memories of being corrected on this in the past, but I can't be sure.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph commented|
|-|-|

I agree that your second example reads more cleanly.  Unfortunately the test is incorrect :).  `.parents()` returns path objects all the way back up to the root, and those have to be determined algorithmically.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph set status to `closed`|
|-|-|

(In [26084]) Add a `parents()` utility method to `FilePath` for iterating all parents of a path back to the filesystem.

Also add this method to `ZipPath`, enumerating all paths back to the root of the archive.

Author: glyph
Reviewer: jml
Fixes [#3588](https://github.com/twisted/twisted/issues/3588)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
